### PR TITLE
Integrate option_map in command

### PR DESCRIPTION
**Pull request by woelke**
In this PR we overhaul `command`and replace `caf::extract_opts()` with `option_map`.

### CODE REVIEW

File: libvast/src/command.cpp [hunk 1]
```diff
@@ -32,29 +32,32 @@ command::~command() {
...
+      break;
+    case option_declaration_set::parse_state::not_an_option:
+      if (position == end)
+        has_subcommand = false;
+      has_subcommand = true;
+      break;
+  }
+  // Check for help option.
+  if (get_or<boolean>(options, "help", false)) {
+    std::cerr << usage() << std::endl;;
```
Comment (woelke) [inline libvast/src/command.cpp]: double `;;`

File: libvast/src/command.cpp [hunk 2]
```diff
@@ -32,29 +32,32 @@ command::~command() {
...
+    default:
+      std::cerr << parse_error(state, position, begin, end) << std::endl;
+      return EXIT_FAILURE;
+    case option_declaration_set::parse_state::successful:
+      has_subcommand = false;
+      break;
+    case option_declaration_set::parse_state::not_an_option:
+      if (position == end)
+        has_subcommand = false;
+      has_subcommand = true;
```
Comment (Neverlord) [inline libvast/src/command.cpp]: You unconditionally set this to `true`. Did you miss an `else` here? In that case, there's no need to branch anyways: `has_subcommand = position != end`.

File: libvast/src/command.cpp [hunk 3]
```diff
@@ -105,53 +107,61 @@ std::string command::full_name() {
...
+                                 argument_iterator error_position,
+                                 argument_iterator begin,
+                                 argument_iterator end) const {
+  std::stringstream result;
+  result << "Failed to parse command" << "\n";
+  result << "  Command: " << name() << " " << detail::join(begin, end, " ")
+         << "\n";
+  result << "  Description: " << to_string(state) << "\n";
+  if (error_position != end)
+    result << "  Position: " << *error_position << "\n";
```
Comment (Neverlord) [inline libvast/src/command.cpp]: This prints the string starting at the position. Is this intentional? I would expect the position to print `distance(begin, position)`.

File: libvast/src/command.cpp [hunk 4]
```diff
@@ -105,53 +107,61 @@ std::string command::full_name() {
...
+  result << "\n" << usage();
+  return result.str();
+}
+
+std::string command::subcommand_error(argument_iterator error_position,
+                                      argument_iterator,
+                                      argument_iterator end) const {
+  std::stringstream result;
+  result << "No such command: " << full_name();
+  if (error_position != end)
```
Comment (Neverlord) [inline libvast/src/command.cpp]: Say I invoke `export`, then this function prints `No such command: export\n\n<help text>`.

To me that's a bit confusing, because `export` *is* a command. The problem is that no subcommand was specified.

File: libvast/src/option_declaration_set.cpp [hunk 5]
```diff
@@ -291,13 +297,34 @@ option_declaration_set::parse(option_map& xs, argument_iterator begin,
...
+constexpr const char* parser_state_names[] = {
+  "successful",
+  "option already exists",
+  "not an option",
+  "name not declared",
+  "argument passed but not declared",
+  "argument declared but not passed",
+  "failed to parse argument",
+  "type not parsebale",
+  "in progress"
```
Comment (Neverlord) [inline libvast/src/option_declaration_set.cpp]: For future reference: a trailing `,` after the last element makes it easier to add entries later and results in a smaller Git diff. The same is true for enums.

File: libvast/src/option_declaration_set.cpp [hunk 6]
```diff
@@ -291,13 +297,34 @@ option_declaration_set::parse(option_map& xs, argument_iterator begin,
...
+};
+
+} // namespace anonymous
+
+const char* to_string(option_declaration_set::parse_state x) {
+  if (x >= option_declaration_set::parse_state::last_state)
```
Comment (Neverlord) [inline libvast/src/option_declaration_set.cpp]: `parse_state` is an enum value that we never receive from external user input, right? In that case I would put an assert here instead of performing a bound check on every lookup.

File: libvast/test/command.cpp [hunk 7]
```diff
@@ -111,41 +108,20 @@ TEST(full name) {
...
   auto cmd1 = root.add<foo>("foo");
-  auto cmd2 = cmd1->add<bar>("bar");
+  cmd1->add<bar>("bar");
   exec("foo -v 42 bar -o 123");
-  CHECK_EQUAL(cmd1->value, 42);
-  CHECK_EQUAL(cmd2->other_value, 123);
-  CHECK_EQUAL(caf::deep_to_string(options),
-              R"([("flag", false), ("other-value", 123), ("value", 42)])");
+  CHECK_EQUAL(get_or(options, "flag", false), false);
+  CHECK_EQUAL(get_or<integer>(options, "value", 0), 42);
```
Comment (Neverlord) [inline libvast/test/command.cpp]: Shouldn't that work without the explicit `<integer>`?

File: libvast/test/option_declaration_set.cpp [hunk 8]
```diff
@@ -83,8 +83,8 @@ TEST(data type parsing) {
   check_fail_option(*x, "X");
   MESSAGE("Test string");
   x = decl.find("string");
-  check_option(*x, "\"2\"", "2");
```
Comment (Neverlord) [inline libvast/test/option_declaration_set.cpp]: Removing the quotations defeats the point of the test. Is this no longer supported by the current implementation?

File: libvast/test/option_map.cpp [hunk 9]
```diff
@@ -99,25 +99,25 @@ TEST(cli parsing) {
   check_option<std::string>(opts, "string", "foo");
   check_fail_option<std::string>(opts, "not-contained");
   MESSAGE("Test long names");
-  args = split("--boolean --integer=42 --string=\"test\"");
```
Comment (Neverlord) [inline libvast/test/option_map.cpp]: Same question here. The option should be able to cope with quoted strings.

File: libvast/vast/command.hpp [hunk 10]
```diff
@@ -87,7 +84,7 @@ class command {
     return is_root() ? *this : parent_->root();
   }
 
-  inline std::string_view name() const noexcept {
+  std::string_view name() const noexcept {
```
Comment (Neverlord) [inline libvast/vast/command.hpp]: Why did you remove `inline` but didn't move the implementation to the `.cpp` file? This results in duplicated symbols. Please make it either inline again for internal linkage or move the implementation to the `.cpp` file. I'd prefer `inline` for small functions like this (3 lines or less is usually a good rule of thumb).

File: libvast/vast/command.hpp [hunk 11]
```diff
@@ -105,75 +102,28 @@ class command {
...
-            double,
-            T
-            >::type
-          >::type;
-      cfg_type copy = ref;
-      return std::make_pair(name, caf::config_value{std::move(copy)});
-    });
-  }
+  expected<void> add_opt(std::string_view name,
+                             std::string_view description, data default_value);
```
Comment (Neverlord) [inline libvast/vast/command.hpp]: Indentation is off.

File: libvast/vast/command.hpp [hunk 12]
```diff
@@ -105,75 +102,28 @@ class command {
...
 
 private:
-  /// Separates arguments into the arguments for the current command, the name
-  /// of the subcommand, and the arguments for the subcommand.
-  std::tuple<caf::message, std::string, caf::message>
-  separate_args(const caf::message& args);
+  std::string parse_error(option_declaration_set::parse_state state,
+                    argument_iterator error_position, argument_iterator begin,
```
Comment (Neverlord) [inline libvast/vast/command.hpp]: Indentation is off.

File: libvast/vast/option_declaration_set.hpp [hunk 13]
```diff
@@ -35,13 +35,14 @@ class option_declaration_set {
...
-    name_not_declartion,
+    not_an_option,
+    name_not_declared,
     arg_passed_but_not_declared,
     arg_declared_but_not_passed,
     failed_to_parse_argument,
     type_not_parsebale,
-    in_progress
+    in_progress,
+    last_state
```
Comment (Neverlord) [inline libvast/vast/option_declaration_set.hpp]: For future code: please leave a trailing `,` (see previous comment).

File: libvast/src/command.cpp (hunk 3 above)
Comment (woelke) [reply libvast/src/command.cpp]: Yes, this is intentional. I think as a user you don't want to count the arguments to see where the error occurred. And `distance(begin, position)` might result in a misleading number because `begin` points to the current sub-command that might be misinterpreted by the user.

I would keep the output at this point and clean it in a separate PR that beautifies the overall output.

File: libvast/src/command.cpp (hunk 4 above)
Comment (woelke) [reply libvast/src/command.cpp]: ok, I see the `subcommand_error()` is a misleading function name.  It is called when the sub-command `export` is not found.  In all other cases `parse_error()` is called. I can rename it to `subcommand_not_found()`.

File: libvast/src/option_declaration_set.cpp (hunk 5 above)
Comment (woelke) [reply libvast/src/option_declaration_set.cpp]: ok

File: libvast/test/command.cpp (hunk 7 above)
Comment (woelke) [reply libvast/test/command.cpp]: You are right, `<integer>` is redundant.

File: libvast/src/command.cpp (hunk 4 above)
Comment (Neverlord) [reply libvast/src/command.cpp]: If you rename this to `subcommand_not_found` than `if (error_position != end)` should be an `ASSERT` instead.

File: libvast/test/option_declaration_set.cpp (hunk 8 above)
Comment (woelke) [reply libvast/test/option_declaration_set.cpp]: Ok, you are right. When we do not use the command line we need the VAST parser for unescaping.
I will add your code snippet.

File: libvast/src/command.cpp (hunk 4 above)
Comment (woelke) [reply libvast/src/command.cpp]: In this case the following typo in command `./vast imbort pcap < ...` would cause an abort of VAST because the position of the parser is at `imbort`which is not valid command.

File: libvast/vast/command.hpp [hunk 14]
```diff
@@ -124,7 +124,7 @@ class command {
                           argument_iterator error_position,
                           argument_iterator begin, argument_iterator end) const;
 
-  std::string subcommand_error(argument_iterator error_position,
+  std::string subcommand_not_found(argument_iterator error_position,
```
Comment (Neverlord) [inline libvast/vast/command.hpp]: Indentation is off.

This function checks for `if (error_position != end)`. Looking at the name this should be a (documented!) precondition instead and only checked via assertion.

File: libvast/src/command.cpp [hunk 15]
```diff
@@ -147,7 +147,7 @@ std::string command::parse_error(option_declaration_set::parse_state state,
   return result.str();
 }
 
-std::string command::subcommand_error(argument_iterator error_position,
+std::string command::subcommand_not_found(argument_iterator error_position,
```
Comment (Neverlord) [inline libvast/src/command.cpp]: Indentation is off.

Comment (woelke) [reply libvast/src/command.cpp]: Arg! And fixed.

File: libvast/vast/command.hpp (hunk 14 above)
Comment (woelke) [reply libvast/vast/command.hpp]: Oh for some reason I overlooked this comment, sorry. So `subcommand_not_found` is still not a good name because it should only create a string which describes what went wrong.
What about `unkown_subcommand_error()` that follows the style of `parse_error`.
